sstar/stats.py

```diff
@@ -0,0 +1,277 @@
+    iv = np.ones((hap_num, 1))
+    counts = tgt_gt*np.matmul(tgt_gt, iv)
+    spectra = np.array([np.bincount(np.array(counts[:,idx] > 0).astype('int8'), minlength=hap_num+1) for idx in range(hap_num)])
+    # ArchIE does not count non-segragating sites
+    spectra[:,0] = 0
+
+    return spectra
+
+
+def cal_n_ton_v2(tgt_gt, ind_num):
+    """
+    Description:
+        Alternative version for calculating individual frequency spetra for haplotypes.
+
+    Arguments:
+        tgt_gt numpy.ndarray: Genotype matrix from the target population.
+        ind_num int: number of individuals 
+
+    Returns:
+        spectra numpy.ndarray: Individual frequency spectra for haplotypes.
+    """
+
+    mut_num, hap_num = tgt_gt.shape
+    max_length = ind_num * 2 + 1
+
+    spectra = []
+    ih_counts = []
+    for ih in range(hap_num):
+        counts = []
+        for elem in tgt_gt:
+            if elem[ih] == 1:
+                counts.append(np.sum(np.array(elem) >0).astype(int))
+
+        spectrum = np.bincount(counts, minlength=hap_num+1)
+  
+        spectra.append(spectrum)
+
+    spectra_length = {len(e) for e in spectra}
+
+    spectra = np.asarray([np.pad(a[0:max_length], (0, max_length - len(a[0:max_length])), 'constant', constant_values=0) for a in spectra])
+
+    return spectra
+
+
+def cal_ref_dist(ref_gt, tgt_gt):
+    """
+    Description:
+        Calculates Euclidean distances between the reference and target populations.
+
+    Arguments:
+        ref_gt numpy.ndarray: Genotype matrix from the reference population.
+        tgt_gt numpy.ndarray: Genotype matrix from the target population.
+
+    Returns:
+        dist_min numpy.ndarray: Minimum distances to the reference population.
+    """
+    ref_dist = distance_matrix(np.transpose(tgt_gt), np.transpose(ref_gt))
+    dist_min = np.min(ref_dist, axis=1)
+
+    return dist_min
+
+
+def cal_tgt_dist(tgt_gt):
+    """
+    Description:
+        Calculates Euclidean distances within the target population.
+
+    Arguments:
+        tgt_gt numpy.ndarray: Genotype matrix from the target population.
+
+    Returns:
+        tgt_dist numpy.ndarray: Pairwise distances between haplotypes from the target population.
+        dist_mean numpy.ndarray: Mean of the pairwise distances between haplotypes from the target population.
+        dist_var numpy.ndarray: Variation of the pairwise distances between haplotypes from the target population.
+        dist_skew numpy.ndarray: Skew of the pairwise distances between haplotypes from the target population.
+        dist_kurtosis numpy.ndarray: Kurtosis of the pairwise distances between haplotypes from the target population.
+    """
+    tgt_dist = distance_matrix(np.transpose(tgt_gt), np.transpose(tgt_gt))
+    # ArchIE sorts tgt_dist
+    tgt_dist.sort()
+    dist_mean = np.mean(tgt_dist, axis=1)
+    dist_var = np.var(tgt_dist, axis=1)
+    dist_skew = sps.skew(tgt_dist, axis=1)
+    dist_kurtosis = sps.kurtosis(tgt_dist, axis=1)
```

Comment:
`skew` and `kurtosis` may be `np.nan` occasionally, we could replace `nan` with 0